### PR TITLE
multivalue hnsw - derive hnsw single and multi from base hnsw

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

This PR implements the multi-value HNSW algorithm. it splits the `hnsw.h` file into `hnsw.h` for the base class and `hnsw_single.h` for the single (regular) case.

**Main objects this PR modified**
1. HNSW index file is now the base class, and hnsw_single and hnsw_multi derive from it.
2. improved the size estimation for HNSW (and modified to support multi-index)
3. "prefer ad-hoc" is changed to check the ratio between the sub-set size and the label count.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

**Review by alonre24 — COMMENTED** [review]
Nice! small comments

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 1]
```diff
@@ -232,8 +232,13 @@ BruteForceIndex<DataType, DistType>::topKQuery(const void *queryBlob, size_t k,
 
     VecSimQueryResult_List rl = {0};
     void *timeoutCtx = queryParams ? queryParams->timeoutCtx : NULL;
-
     this->last_mode = STANDARD_KNN;
+
+    if (0 == k) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/brute_force/brute_force.h:237]: I'll add a test. I think it is a good practice to have the literal part of the comparison on the left side so it is not mistaken with asserting the value.

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp [hunk 2]
```diff
@@ -37,25 +51,39 @@ size_t EstimateInitialSize(const HNSWParams *params) {
...
-    // For every new vector, a new node of size 24 is allocated in a bucket of the hash table.
-    size_t size_label_lookup_node =
-        24 + sizeof(size_t); // 24 + VecSimAllocator::allocation_header_size
+    size_t size_label_lookup_node;
+    if (params->multi) {
+        // For each new insertion (of a new label), we add a new node to the label_lookup_ map,
+        // and a new element to the vector in the map. These two allocations both results in a new
+        // allocation and therefore another VecSimAllocator::allocation_header_size.
+        size_label_lookup_node =
+            sizeof(vecsim_stl::unordered_map<labelType, vecsim_stl::vector<idType>>::value_type) +
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_factory.cpp:77]: my tests showed that this value predicts more accurately the memory consumption

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 3]
```diff
@@ -412,26 +427,28 @@ DistType HNSWIndex<DataType, DistType>::processCandidate(
...
 
-            top_candidates.emplace(dist1, candidate_id);
+            emplaceToHeap(top_candidates, dist1, candidate_id);
 
             if (top_candidates.size() > ef)
                 top_candidates.pop();
 
-            if (!top_candidates.empty())
-                lowerBound = top_candidates.top().first;
+            lowerBound = std::max(top_candidates.top().first, dist1);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:450]: Because we return early on `k == 0`, and that's the only case when ef can be 0, the heap must contain at least one element from the insertion two lines above

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 4]
```diff
@@ -952,8 +968,8 @@ HNSWIndex<DataType, DistType>::HNSWIndex(const HNSWParams *params,
     level_generator_.seed(random_seed);
 
     // data_level0_memory will look like this:
-    // -----4------ | -----4*M0----------- | ----8------------------| ------32------- | ----8---- |
-    // <links_len>  | <link_1> <link_2>... | <incoming_links_set> |   <data>        |  <label>
+    // | -----4------ | -----4*M0----------- | ----------8----------| --data_size_-- | ----8---- |
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:971]: :pray:

File: tests/unit/test_hnsw.cpp [hunk 5]
```diff
@@ -47,8 +47,6 @@ TEST_F(HNSWTest, hnsw_blob_sanity_test) {
         ASSERT_FALSE(memcmp(v, blob, sizeof(blob)));                                               \
     } while (0)
 
-#define GET_LABEL(id) hnsw_index->getExternalLabel(id)
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_hnsw.cpp:50]: I added it when we had to get the internal HNSW first and then call `getExternalLabel`. Seems redundant now.

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 6]
```diff
@@ -1509,27 +1505,24 @@ VecSimQueryResult_List HNSWIndex<DataType, DistType>::rangeQuery(const void *que
...
     }
 
     // search bottom layer
-    rl.results = searchRangeBottomLayer_WithTimeout(bottom_layer_ep, query_data, this->epsilon_,
-                                                    radius, timeoutCtx, &rl.code);
+    rl.results = searchRangeBottomLayer_WithTimeout(bottom_layer_ep, query_data, epsilon, radius,
+                                                    timeoutCtx, &rl.code);
 
-    // Restore the default epsilon.
-    epsilon_ = originalEpsilon;
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw.h:1532]: why did you remove this?

Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw.h:1532]: because I don't change it in the first place

File: tests/unit/test_bruteforce.cpp [hunk 7]
```diff
@@ -371,6 +371,34 @@ TEST_F(BruteForceTest, brute_force_vector_search_by_id_test) {
     VecSimIndex_Free(index);
 }
 
+TEST_F(BruteForceTest, search_for_nothing) {
```
Comment (alonre24) [inline tests/unit/test_bruteforce.cpp:374]: Consider adding this check to an exiting basic test such as `brute_force_vector_search_test_l2` maybe, to reduce the inflation of the test files...

File: tests/unit/test_hnsw.cpp [hunk 8]
```diff
@@ -71,33 +69,62 @@ TEST_F(HNSWTest, hnsw_blob_sanity_test) {
...
     ASSERT_HNSW_BLOB_EQ(1, d);
-    ASSERT_EQ(GET_LABEL(0), 46);
-    ASSERT_EQ(GET_LABEL(1), 42);
+    ASSERT_EQ(hnsw_index->getExternalLabel(0), 46);
+    ASSERT_EQ(hnsw_index->getExternalLabel(1), 42);
+
+    VecSimIndex_Free(index);
+}
+
+TEST_F(HNSWTest, search_for_nothing) {
```
Comment (alonre24) [inline tests/unit/test_hnsw.cpp:100]: same here (maybe move to a basic search test function)

File: tests/unit/test_hnsw_multi.cpp [hunk 9]
```diff
@@ -0,0 +1,1989 @@
...
+    ASSERT_EQ(HYBRID_BATCHES_TO_ADHOC_BF, info.hnswInfo.last_mode);
+    compareHNSWIndexInfoToIterator(info, infoIter);
+    VecSimInfoIterator_Free(infoIter);
+
+    VecSimIndex_Free(index);
+}
+
+TEST_F(HNSWMultiTest, preferAdHocOptimization) {
+    // Save the expected result for every combination that represent a different leaf in the tree.
+    // map: [k, index_size, dim, M, r] -> res
```
Comment (alonre24) [inline tests/unit/test_hnsw_multi.cpp:572]: Update the comment about the map. Also, you can call explicitly `VecSimIndex_PreferAdHocSearch` for single to set the expected values, or at least document how you set these values.

**Review by alonre24 — DISMISSED** [review]

File: tests/unit/test_hnsw_multi.cpp [hunk 10]
```diff
@@ -0,0 +1,1989 @@
...
+        float expected_score =
+            1.0f -
+            ((first_coordinate + (float)dim - 1.0f) /
+             (sqrtf((float)dim) * sqrtf((float)(dim - 1) + first_coordinate * first_coordinate)));
+        // Verify that abs difference between the actual and expected score is at most 1/10^6.
+        ASSERT_NEAR(score, expected_score, 1e-5);
+    };
+    runTopKSearchTest(index, query, 10, verify_res);
+
+    // Test with batch iterator.
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_multi.cpp:1852]: should we split it?

**Review by DvirDukhan — APPROVED** [review]
